### PR TITLE
add semicolons, new lines, and formatting to chapel-syntax-printer

### PR DESCRIPTION
**Pull request by arezaii**
This PR updates the Chapel syntax printer to include semicolons,
new lines, indentation and basic support for including parentheses
based on operator precedence. 

Future work remains on operator precedence support to handle cases
where operations on different types may cause underflow/overflow,
as pointed out by @cassella at https://github.com/chapel-lang/chapel/pull/19902#issuecomment-1140144485:
```
var a: int;
var b, c: int(8);
```
> means that a+(b+c) does need its parens due to overflow/underflow.

reviewed by @dlongnecke-cray - thanks!

TESTING:

- [x] paratest

Signed-off-by: arezaii <ahmad.rezaii@hpe.com>

### CODE REVIEW

**Comment by cassella** [comment]
BTW, #17169 did something similar wrt parenthesizing expressions for chpldoc emitting initialization values and argument default values.  Note it's more than just precedence -- `+` and `-` have the same precedence but `a+(b+c)` doesn't need its parens, while `a-(b-c)` does.  Also, unary `+` and `-` have different precedence than the binary operators.

**Comment by arezaii** [comment]
Thanks @cassella! I implemented a very rudimentary, incomplete way of doing this, so I greatly appreciate your past contribution and you pointing it out to me. 😸

**Comment by arezaii** [comment]
I made https://github.com/chapel-lang/chapel/pull/19902 to begin integrating the work from https://github.com/chapel-lang/chapel/pull/17169 into these changes.

File: compiler/dyno/lib/uast/chpl-syntax-printer.cpp [hunk 1]
```diff
@@ -712,47 +880,57 @@ struct ChplSyntaxVisitor {
...
-    printBlockWithStyle(node->blockStyle(), node->stmts(), "do ");
+    printBlockWithStyle(node->blockStyle(), node->stmts(), "do ", ";", true);
   }
 
   void visit(const Module* node) {
     if (node->visibility() != Decl::Visibility::DEFAULT_VISIBILITY) {
       ss_ << kindToString(node->visibility()) << " ";
     }
-    if (node->kind() != Module::Kind::DEFAULT_MODULE_KIND ) {
+    if (node->kind() != Module::Kind::DEFAULT_MODULE_KIND && node->kind() != Module::Kind::IMPLICIT) {
```
Comment (dlongnecke-cray) [inline compiler/dyno/lib/uast/chpl-syntax-printer.cpp:897]: My line width senses are tingling, let's make sure to keep at 80 cols.

File: compiler/dyno/lib/uast/chpl-syntax-printer.cpp [hunk 2]
```diff
@@ -712,47 +880,57 @@ struct ChplSyntaxVisitor {
...
+    std::string kind;
+    for (auto decl : node->decls()) {
+      if (!decl->isComment()) {
+        if (decl->isVariable() ) {
+          if (decl->toVariable()->isConfig())
+            isConfig = true;
+          kind = kindToString((IntentList)decl->toVariable()->kind());
+        }
+        else if (decl->isTupleDecl()) {
+          // TODO: can tuple decls be configs?
```
Comment (dlongnecke-cray) [inline compiler/dyno/lib/uast/chpl-syntax-printer.cpp:916]: I think they definitely can be and we can parse that (IIRC?), it's just we don't currently support setting them and so don't support them as a result. They require a recursive sort of logic we just haven't gotten around to doing.

File: compiler/dyno/lib/uast/chpl-syntax-printer.cpp [hunk 3]
```diff
@@ -1024,14 +1229,14 @@ struct ChplSyntaxVisitor {
       interpose(node->caseExprs(), ", ");
       ss_ << " ";
     }
-    printBlockWithStyle(node->blockStyle(), node->stmts(), "do ");
+    printBlockWithStyle(node->blockStyle(), node->stmts(), "do ", ";", true);
   }
 
   void visit(const While* node) {
     ss_ << "while ";
     printChapelSyntax(ss_, node->condition());
```
Comment (dlongnecke-cray) [inline compiler/dyno/lib/uast/chpl-syntax-printer.cpp:1237]: Do we wanna switch this to `printAst`?

File: compiler/dyno/test/uast/testStringify.cpp [hunk 4]
```diff
@@ -436,9 +455,12 @@ static void test4(Parser* parser) {
 
 static void test5(Parser* parser) {
   // unsure how best to set this up for testing
+  // probably need to parse an input to a node, and send that node into a function
```
Comment (dlongnecke-cray) [inline compiler/dyno/test/uast/testStringify.cpp:458]: I think the best thing to do is just to map a string to a file using `setFileContents`. Then you parse and syntaxify it, and can print that to a string stream. Then you just assert that the stream contents and the file contents are the same.

```cpp
auto path = UniqueString::get(context, "test5.chpl");
std::string contents = R""""(
module Foobar {}
)"""";
setFileText(context, path, contents);
auto& br = parseFile(context, path);
assert(!br.numErrors());
auto mod = br.singleModule();
assert(mod);
std::ostringstream ss;
printChapelSyntax(ss, mod);
assert(ss.str() == fileText(context, path).text());
```

Ah, but I just read the macros and you are already doing something similar. I'm not sure what the problem is here, feel free to poke me to talk about it!

File: compiler/dyno/lib/uast/chpl-syntax-printer.cpp [hunk 5]
```diff
@@ -767,35 +945,75 @@ struct ChplSyntaxVisitor {
...
     ss_ << " ";
-    printBlockWithStyle(node->blockStyle(), node->stmts(), "do ");
+    printBlockWithStyle(node->blockStyle(), node->stmts(), "do ", ";", true);
   }
 
   void visit(const OpCall* node) {
-    // TODO: Where do parens come in?
-    // ex: !(this && that)
-    // is different than !this && that
+    bool needsParens = false;
```
Comment (dlongnecke-cray) [inline compiler/dyno/lib/uast/chpl-syntax-printer.cpp:952]: I might make two helper functions e.g. `printBinaryOp` and `printUnaryOp` to reduce the branching a bit.

**Review by dlongnecke-cray — APPROVED** [review]
Looks good, mainly just nits. I see a couple places where `printChapelSyntax` is called, and I think we probably want to switch those to calls to `printAst`?

It looks like you're running into a problem with `test5`, feel free to talk to me about it if you need.

File: compiler/dyno/lib/uast/chpl-syntax-printer.cpp [hunk 6]
```diff
@@ -123,15 +123,85 @@ static std::string pragmaFlagsToString(const Decl* node) {
...
+  if (node->index() && node->stmt(0)->isConditional()) return false;
+  return true;
+}
 
+// TODO: Attributes
 
 struct ChplSyntaxVisitor {
   std::stringstream ss_;
+  int indentDepth_ = 0;
+  const int spacesPerIndentDepth_ = 2;
```
Comment (dlongnecke-cray) [inline compiler/dyno/lib/uast/chpl-syntax-printer.cpp:141]: Can make this static so we don't need to reserve space for it

**Review by dlongnecke-cray — APPROVED** [review]
Nice cleanup. This is great progress!

**Pull request merged**